include/libdnf-cli/progressbar/progress_bar.hpp

```diff
@@ -0,0 +1,150 @@
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+
+#ifndef LIBDNF_CLI_PROGRESSBAR_PROGRESS_BAR_HPP
+#define LIBDNF_CLI_PROGRESSBAR_PROGRESS_BAR_HPP
+
+
+#include <chrono>
+#include <string>
+#include <vector>
+
+
+namespace libdnf::cli::progressbar {
+
+
+enum class ProgressBarState : int {
+    READY,    // no color
+    STARTED,  // no color
+    SUCCESS,  // green
+    WARNING,  // yellow
+    ERROR,    // red
+};
+
+
+enum class MessageType : int {
+    INFO,     // no color
+    SUCCESS,  // green
+    WARNING,  // yellow
+    ERROR,    // red
+};
+
+
+
+
+
+class ProgressBar {
+public:
+    using Message = std::pair<MessageType, std::string>;
+
+    explicit ProgressBar(int64_t total_ticks);
+    explicit ProgressBar(int64_t total_ticks, const std::string & description);
+    virtual ~ProgressBar() = default;
+
+    // ticks
+    int64_t get_ticks() { return ticks; }
+    void set_ticks(int64_t value);
+    void add_ticks(int64_t value);
+
+    // total ticks
+    int64_t get_total_ticks() { return total_ticks; }
+    void set_total_ticks(int64_t value);
+    //void add_total_ticks(int64_t value);
+
+    // number
+    int32_t get_number() { return number; }
+    void set_number(int32_t value) { number = value; }
+
+    // total
+    int32_t get_total() { return total; }
+    void set_total(int32_t value) { total = value; }
+
+    // workflow
+    void start();
+    ProgressBarState get_state() { return state; }
+    void set_state(ProgressBarState value) { update(); state = value; }
+    bool is_finished() { return get_state() != ProgressBarState::READY && get_state() != ProgressBarState::STARTED; }
+    bool is_failed() { return get_state() == ProgressBarState::ERROR || get_state() == ProgressBarState::WARNING; }
+
+    // description
+    std::string get_description() const { return description; }
+    void set_description(const std::string & value) { description = value; }
+
+    // messages
+    void add_message(MessageType type, const std::string & message) { messages.push_back({type, message}); }
+    const std::vector<Message> & get_messages() { return messages; }
+
+    // auto-finish feature; turn off if you want to handle state manually
+    bool get_auto_finish() { return auto_finish; }
+    void set_auto_finish(bool value) { auto_finish = value; }
+
+    // stats
+    void update();
+    int32_t get_percent_done() { return percent_done; }
```

Comment:
This type of getters can be marked as `const noexcept`.

Reply:
fixed